### PR TITLE
Building ARM binaries in the pipeline

### PR DESCRIPTION
**Pull request by serban-nicusor-toptal**
## Release notes

Added build stage for ARM in Jenkinsfile.  
I have added a new EC2 Fleet for ARM with 0 instances so it will spin up only when we need it, this way we can now easily build ARM for other projects too.  

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -272,6 +272,34 @@ pipeline {
...
+                            dune build @install --profile static
+                        """)
+
+                        echo runShell("""
+                            eval \$(opam env)
+                            time dune runtest --verbose
+                        """)
+
+                        sh "mkdir -p bin && mv `find _build -name stanc.exe` bin/arm-stanc"
+                        sh "mv _build/default/src/stan2tfp/stan2tfp.exe bin/arm-stan2tfp"
```
Comment (rok-cesnovar) [inline Jenkinsfile:296]:
```suggestion
                        sh "mkdir -p bin && mv `find _build -name stanc.exe` bin/linux-arm-stanc"
                        sh "mv _build/default/src/stan2tfp/stan2tfp.exe bin/linux-arm-stan2tfp"
```

File: Jenkinsfile [hunk 2]
```diff
@@ -272,6 +272,34 @@ pipeline {
...
+
+                        stash name:'arm-exe', includes:'bin/*'
```
Comment (rok-cesnovar) [inline Jenkinsfile:298]:
```suggestion
                        stash name:'linux-arm-exe', includes:'bin/*'
```

File: Jenkinsfile [hunk 3]
```diff
@@ -313,6 +341,7 @@ pipeline {
                 unstash 'windows-exe'
                 unstash 'linux-exe'
                 unstash 'mac-exe'
+                unstash 'arm-exe'
```
Comment (rok-cesnovar) [inline Jenkinsfile:344]:
```suggestion
                unstash 'linux-arm-exe'
```

**Review by rok-cesnovar — CHANGES_REQUESTED** [review]
Looks great, maybe we rename this to linux-arm?

**Comment by serban-nicusor-toptal** [comment]
Thanks @rok-cesnovar ! I've renamed that locally but will wait before pushing to see if the pipeline is working as expected.
I have added a new `EC2 Fleet` for ARM with `0` instances so it will spin up only when we need it, this way we can now easily build ARM for other projects too.
I've seen that the [binaries workflow](https://github.com/stan-dev/stanc3/blob/master/.github/workflows/binaries.yml) is used so I will make that build ARM too if that's okay with you.

**Comment by rok-cesnovar** [comment]
> I've seen that the binaries workflow is used so I will make that build ARM too if that's okay with you.

Definitely!

**Comment by serban-nicusor-toptal** [comment]
Argh looks like Github Actions [do not offer](https://docs.github.com/en/actions/reference/specifications-for-github-hosted-runners#about-github-hosted-runners) ARM based VMs as a service, only if we self-host an ARM based VM and use it for builds.
I can make the Jenkins ARM instance to self-register to GitHub  ( to avoid having one 24/7 ) when it boots but this is only working if we start a Jenkins job first, if it's the other way around GitHub won't have a machine available. Now it's a matter of usability I guess, if we only need it for releases it's enough in Jenkins, if we want to have arm in gh actions then we should go for it. I'm up for any that fits us best.
What do you think @rok-cesnovar ?

**Comment by rok-cesnovar** [comment]
Ah yes, they do not. 

In that case its just Jenkins for ARM on releases.The point of the GHA builder is that anyone can clone the repo and build the binaries for themselves so those cant rely on our instances/Jenkins.

**Comment by serban-nicusor-toptal** [comment]
Hey Rok, any idea why it's failing [here](https://jenkins.mc-stan.org/blue/organizations/jenkins/stanc3/detail/PR-832/13/pipeline/) ?

**Comment by rok-cesnovar** [comment]
This seems to be caused by slight numerical differences (maybe due to different CPU architecture). Just comment out `eval $(opam env) time dune runtest --profile static --verbose` as it isnt really needed for building binaries. We can make this data gen resistant to this numerical differences at a later point.

**Review by rok-cesnovar — APPROVED** [review]

**Comment by rok-cesnovar** [comment]
Oh right, that would be better. So you now have to spawn the arm instance manually so it builds? Or it just takes awhile to spawn?

**Comment by serban-nicusor-toptal** [comment]
It should spawn by itself when needed if everything works as expected.

**Comment by rok-cesnovar** [comment]
Cool, then lets apply the fix you suggest.

**Comment by serban-nicusor-toptal** [comment]
For some reason it's taking way longer than expected:
image
I need to find a fix for this later, looks like it doesn't like our region :(

**Pull request merged**